### ISSUE TITLE
`IpAddr::from_str` is unnecessarily slow for IPv6 addresses

### ISSUE THREAD

**Issue by jyn514** [rust-lang/rust#94825]
`from_str` is implemented like so: https://github.com/rust-lang/rust/blob/5f4e0677190b82e61dc507e3e72caf89da8e5e28/library/std/src/net/parser.rs#L226-L229
This unnecessarily penalizes ipv6 addresses, because the parser will always do a full linear scan of the input before it even starts to parse the address. Instead, it could use the knowledge that IPv4 addresses are at most 15 bytes long to skip directly to `read_ipv6_addr` for longer strings.

@rustbot label: +I-slow +T-libs +C-enhancement

**Comment by jyn514** [comment]
> IPv4 addresses are at most 15 bytes long

This is true because octal addresses were banned in https://github.com/rust-lang/rust/pull/83652.

**Comment by equt** [comment]
I do agree that could benefit the `read_ip_addr` and maybe also the `read_socket_addr`. But I believe that

> the parser will always do a full linear scan of the input

is not true.

If I didn't get it wrong, the current parser will realize the input is an IPv6 after parsing at most four bytes. And if we're going to use the input size as a factor to guess, something like `0000::1` that doesn't meet will always cause the parser to backtrace.

**Comment by mbartlett21** [comment]
> the current parser will realize the input is an IPv6 after parsing at most four bytes

That seems correct

https://github.com/rust-lang/rust/blob/21b0325c68421b00c6c91055ac330bd5ffe1ea6b/library/std/src/net/parser.rs#L145-L148

**Comment by Nilstrieb** [comment]
There are probably still further performance improvements that could be made, but unless someone has a real-world use case for parsing that many IPv6 addresses, this seems unnecessary.

**Comment by saethlin** [comment]
> unless someone has a real-world use case for parsing that many IPv6 addresses

My employer maintains an ETL process that mostly chomps through netflow data, spending a nontrivial amount of cycles parsing IP addresses depending on configuration.

I don't particularly see any reason to close this; if we can have super-fast IP address parsing, we should.

**Comment by niklasf** [comment]
#158053 makes some progress without doing anything too arcane.

**Comment by tbu-** [comment]
Since the initial premise of this issue is wrong

> This unnecessarily penalizes ipv6 addresses, because the parser will always do a full linear scan of the input before it even starts to parse the address.

and there have been performance improvements with #158053, I think this issue should be closed unless someone benchmarks to see that "`IpAddr::from_str` is unnecessarily slow for IPv6 addresses".

**Comment by jyn514** [comment]
seems reasonable. i no longer work at the company where this was a problem; i'll let @saethlin follow-up if he still sees it in benchmarks.
